coredns: add support for plugins

bct opened the pull request:
## Description of changes

Solves https://github.com/NixOS/nixpkgs/issues/146603

Usage:

```
coredns-fanout = pkgs.coredns.override {
  externalPlugins = [
    {name = "fanout"; repo = "github.com/networkservicemesh/fanout"; version = "v1.9.1";}
  ];
  vendorHash = "<SRI hash>";
};
```

This PR is based on https://github.com/NixOS/nixpkgs/pull/205649 . I've made 3 changes:

- rebased it on master
- added `go mod vendor` to the end of `modBuildPhase`
- modified the `externalPlugins` argument to take the plugin name, repo and version
  - this allows plugins to be pinned, which was a request made on the original PR
- dropped the list of official upstream plugins (`external-plugins.nix`)
  - this was convenient but not necessary. removing it means it doesn't need to be kept up to date.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

de11n commented on nixos/doc/manual/release-notes/rl-2311.section.md:
````diff
@@ -306,6 +306,21 @@ The module update takes care of the new config syntax and the data itself (user
+  };
+};
+```
+To get the necessary sha256 hash, set `vendorSha256 = lib.fakeHash;`. The build will fail and produce your sha256 hash in the error message. You should copy that hash and use it for your `vendorSha256`.
````
Recommend `vendorHash = "";` because it's much easier and doesn't require dependency on `lib`.

de11n commented on pkgs/servers/dns/coredns/default.nix:
```diff
@@ -16,12 +23,40 @@ buildGoModule rec {
+
+  # Copy over the lockfiles as well, because the source
+  # doesn't contain it. The fixed-output derivation is
+  # probably not reproducible anyway.
```
We should try our best to *make* it reproducible. Maybe the comment is out of date, now that we pin the versions of plugins?

de11n reviewed (changes requested):
Great work! I took a stab at this a while ago and came up with something very similar. I have a few small comments but largely LGTM.

bct commented:
I made the suggested changes.

de11n commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -306,6 +306,21 @@ The module update takes care of the new config syntax and the data itself (user
+    externalPlugins = [
+      {name = "fanout"; repo = "github.com/networkservicemesh/fanout"; version = "v1.9.1";}
+    ];
+    vendorHash = "<sha256 hash>";
```
nit: 
```suggestion
    vendorHash = "<SRI hash>";
```

de11n commented on nixos/doc/manual/release-notes/rl-2311.section.md:
````diff
@@ -306,6 +306,21 @@ The module update takes care of the new config syntax and the data itself (user
+};
+```
+To get the necessary sha256 hash, set `vendorHash = "";`. The build will fail and produce your sha256 hash in the error message. You should copy that hash and use it for your `vendorHash`.
+Please note! If you use this feature, you will not receive updates to CoreDNS when you upgrade your system. To update CoreDNS, you'll have to update `vendorHash`; for that, set `vendorHash = "";` again and follow the steps above.
````
Technically it's a bit more subtle than that. The newer version of `coredns` may attempt to build with the old `vendor`. It may even build successfully if nothing significant has changed. Or it may entirely fail to build. I'm not sure how much of that detail is worth putting in this note however. Just pointing it out.

bct replied:
Yeah, this section was a little too detailed and repetitive IMO. I've streamlined it.

de11n reviewed (changes requested):
Small comments. Overall looking great.

de11n commented on pkgs/servers/dns/coredns/default.nix:
```diff
@@ -16,19 +23,44 @@ buildGoModule rec {
 
+  # Override the go-modules fetcher derivation to fetch plugins
+  modBuildPhase = ''
+    for plugin in ${builtins.toString (attrsToPlugins externalPlugins)}; do echo $plugin >> plugin.cfg; done
```
Extreme nit: You can build the entire string in Nix and then only echo once. But this is fine and matches the loop below it.

de11n reviewed (approved):
I don't have merge bits, but looks great. You may want to squash your commits to make it easier for whoever does merge it.

bct committed:
coredns: allow adding external plugins

Solves https://github.com/NixOS/nixpkgs/issues/146603

CoreDNS has support for plugins that are added at compile time. This
exposes an argument `externalPlugins` that will build coredns with
the specified plugins.

Example:
```
coredns-fanout = pkgs.coredns.override {
  externalPlugins = [
    {name = "fanout"; repo = "github.com/networkservicemesh/fanout"; version = "v1.9.1";}
  ];
  vendorHash = "<SRI hash>";
};
```

bct commented:
Thanks! I've squashed it into a single commit.

bjornfor reviewed (commented):
Commit message typo: `plugins tht are` -> `plugins that are`

KFearsoff reviewed (approved):
LGTM

Regarding the "newer CoreDNS rebuilds with older vendor": do we have a precedent for handling something like this in Nixpkgs? I don't believe we do, and there's no easy way to handle it, right?

bjornfor commented:
When coredns gets updated, and I use externalPlugins, does the build continue with stale vendored code, or does it fail?

bct commented:
My understanding (based on trying the 1.10.1 -> 1.11.0 upgrade and reading https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/go/module.nix ):

The package version is part of the `-go-modules` derivation name. When the version changes `goModules` will be rebuilt. If the vendored dependencies have changed, then the `vendorHash` won't match and the build will fail.

I was worried that it might be possible for a new version to build successfully against old modules if you had a cached `goModules`, but I don't think this can happen since the version is part of the derivation name.

bjornfor commented:
> The package version is part of the -go-modules derivation name. When the version changes goModules will be rebuilt. If the vendored dependencies have changed, then the vendorHash won't match and the build will fail. If the vendored dependencies haven't changed then it will build successfully.

Perfect. Thanks!

bjornfor commented:
Thanks!

evrim commented:
sry to barge in like this but coredns is failing in staging-next with the following, mite be related.

```
Below is the build log of derivation /nix/store/fy12vfg6j6v86j997is0chhnypis75c9-coredns-1.11.0.drv. It was built on ssh://localhost.

unpacking sources
unpacking source archive /nix/store/9g5sy1favi7bawr3ynjdid2ssj0hnwv3-source
source root is source
patching sources
updateAutotoolsGnuConfigScriptsPhase
configuring
building
Building subPackage .
# github.com/quic-go/quic-go/internal/handshake
vendor/github.com/quic-go/quic-go/internal/handshake/crypto_setup.go:362:37: cannot use h.allow0RTT (variable of type bool) as tls.QUICSessionTicketOptions value in argument to h.conn.SendSessionTicket
```

bct commented:
This looks like the same issue: https://github.com/quic-go/quic-go/issues/4021

Possibly this is a result of https://github.com/NixOS/nixpkgs/pull/254878 ?

evrim commented:
So, what mite be the correct approach here; we can just do buildGo119Module to build and it will succeed I guess. How about other options?

bct commented:
coredns 1.11.1 is supposed to have the fix (https://github.com/coredns/coredns/pull/6252). it builds successfully for me on staging-next:

```
diff --git a/pkgs/servers/dns/coredns/default.nix b/pkgs/servers/dns/coredns/default.nix
index 2dcfc538be45..8078018625a2 100644
--- a/pkgs/servers/dns/coredns/default.nix
+++ b/pkgs/servers/dns/coredns/default.nix
@@ -4,7 +4,7 @@
 , fetchFromGitHub
 , installShellFiles
 , externalPlugins ? []
-, vendorHash ? "sha256-TvIswNQ7DL/MtYmMSxXf+VqKHcmzZVZwohOCvRWxBkY="
+, vendorHash ? "sha256-tp22jj6DNnYFQhtAFW2uLo10ty//dyNqIDH2egDgbOw="
 }:

 let
@@ -14,13 +14,13 @@ let
     builtins.map ({name, repo, version}: "${repo}@${version}") attrs;
 in buildGoModule rec {
   pname = "coredns";
-  version = "1.11.0";
+  version = "1.11.1";

   src = fetchFromGitHub {
     owner = "coredns";
     repo = "coredns";
     rev = "v${version}";
-    sha256 = "sha256-Mn8hOsODTlnl6PJaevMcyIKkIx/1Lk2HGA7fSSizR20=";
+    sha256 = "sha256-XZoRN907PXNKV2iMn51H/lt8yPxhPupNfJ49Pymdm9Y=";
   };

   inherit vendorHash;
```

I can put up a PR for this tonight.

evrim commented:
Oh that's great, thnx!

bct commented:
https://github.com/NixOS/nixpkgs/pull/258136
